File: akka-samples/scala-samples/src/main/scala/actors/request/RequestResponse.scala
package actors.request

import akka.actor.typed.{ActorRef, ActorSystem, Behavior, Terminated}
import akka.actor.typed.scaladsl.Behaviors

object RequestResponse {
  def main(args: Array[String]): Unit = {
    ActorSystem[String](Guardian(), "request-response")
  }
}

private object Guardian {
  final case class Request(query: String, replyTo: ActorRef[Response])
  final case class Response(result: String)

  def requester(target: ActorRef[Request]): Behavior[Response] =
    Behaviors.setup[Response] { context =>
      target ! Request("PI", context.self)

      Behaviors.receive[Response] { (context, message) =>
        println(message.result)
        Behaviors.stopped
      }
    }

  def responder(): Behavior[Request] =
    Behaviors.receive[Request] { (context, message) =>
      println(s"Received actors.request: ${message.query}")

      message.replyTo ! Response(s"Your Result: ${Math.PI}")
      Behaviors.stopped
    }

  def apply(): Behavior[String] =
    Behaviors.setup[String] { context =>
      val responder: ActorRef[Request] = context.spawnAnonymous(Guardian.responder())
      val requester: ActorRef[Response] = context.spawnAnonymous(Guardian.requester(responder))
      context.watch(requester)

      Behaviors.receiveSignal {
        case (context, Terminated(ref)) =>
          Behaviors.stopped
      }
    }
}

File: akka-samples/java-samples/src/main/java/actors/request/RequestResponse.java
package actors.request;

import akka.actor.typed.ActorRef;
import akka.actor.typed.ActorSystem;
import akka.actor.typed.Behavior;
import akka.actor.typed.Terminated;
import akka.actor.typed.javadsl.AbstractBehavior;
import akka.actor.typed.javadsl.ActorContext;
import akka.actor.typed.javadsl.Behaviors;
import akka.actor.typed.javadsl.Receive;

public class RequestResponse extends AbstractBehavior<String> {

    @Override
    public Receive<String> createReceive() {
        return newReceiveBuilder().onSignal(Terminated.class, (message) -> Behaviors.stopped()).build();
    }

    public RequestResponse(ActorContext<String> context) {
        super(context);

        ActorRef<Responder.Request> responder = context.spawnAnonymous(Responder.create());
        ActorRef<Requester.Response> requester = context.spawnAnonymous(Requester.create(responder));
        context.watch(requester);
    }

    public static Behavior<String> create() {
        return Behaviors.setup(RequestResponse::new);
    }

    public static void main(String[] args) {
        ActorSystem.create(RequestResponse.create(), "request-response");
    }
}
